### PR TITLE
Make opaque::Encoder append-only and make it infallible

### PR DESCRIPTION
**Pull request by Zoxc**

### CODE REVIEW

File: src/librustc_incremental/persist/save.rs [hunk 1]
```diff
@@ -246,13 +234,12 @@ fn encode_work_product_index(work_products: &FxHashMap<WorkProductId, WorkProduc
         })
         .collect();
 
-    serialized_products.encode(encoder)
+    serialized_products.encode(encoder).ok();
```
Comment (michaelwoerister) [inline src/librustc_incremental/persist/save.rs]: Should this be `.unwrap()` too?

**Comment by michaelwoerister** [comment]
Thanks, @Zoxc! Looking forward to seeing perf results for this.

**Comment by Mark-Simulacrum** [comment]
We'll need to wait for https://github.com/rust-lang/rust/pull/51680 to merge before collecting perf becomes feasible.

**Comment by Zoxc** [comment]
Perf link when done: https://perf.rust-lang.org/compare.html?start=f790780451a724dce9be0ba03ff217955036ff88&end=ec349a9683138a28c5cb45a0492a292c66e64a68&stat=instructions%3Au

**Comment by Zoxc** [comment]
@Mark-Simulacrum A perf run is still wanted here

**Comment by Mark-Simulacrum** [comment]
I've queued the run.

**Comment by michaelwoerister** [comment]
Perf looks awesome! r=me with that one nit fixed.

**Pull request merged**